### PR TITLE
feat: body-transformer plugin enhancement(#10472)

### LINKED ISSUE #10472: feat:  body-transformer enhancement
#### 1、Backend returns gzip not supported
When the backend returns gzip data,  the decoding will be garbled.  i want to fix it

#### 2、Data using x-ww-form-urlencoded format is not supported
When the client sends a request format of x-www-form-urlencoded, Unable to parse parameter

#### 3、Parameter parsing and conversion of get requests are not supported
When the client sends a get request, Unable to parse parameter

### target
I want to add these features which are not supported right now
1. Support backend to return gzip data
2. Support the use of x-www-form-urlencoded format
3. Support the use of get requests

### PR DESCRIPTION
**Pull request by yongxiaodong**
### Description
- Support content-type using x-www-form-urlencoded format
- Supports parsing uri parameters of get requests

Fixes #10472

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

**Comment by yongxiaodong** [comment]
CLI Test / build (ubuntu-20.04, linux_apisix_current_luarocks_in_customed_nginx) (pull_request) seems like it shouldn't fail, can you help me point out the reason?

File: t/plugin/body-transformer.t [hunk 1]
```diff
@@ -894,3 +894,178 @@ location /demo {
...
+                return
+            end
+            ngx.sleep(0.5)
+
+            local core = require("apisix.core")
+            local http = require("resty.http")
+            local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/foobar"
+            local data = {name = "hello", age = 20}
+            local body = ngx.encode_args(data)
+            local opt = {method = "POST", body = body, headers = {["Content-Type"] = "application/x-www-form-urlencoded"}}
```
Comment (shreemaan-abhishek) [inline t/plugin/body-transformer.t:949]: why this body is in plain json and not url encoded?

**Comment by shreemaan-abhishek** [comment]
Do you plan to support gzip as mentioned in the linked issue?

**Comment by yongxiaodong** [comment]
> Do you plan to support gzip as mentioned in the linked issue?

The apisix gateway usually transmits data through the intranet and upstream. I think there are very few usage scenarios of gzip on the intranet, and gzip will greatly reduce the performance of apisix. What do you think?

File: apisix/plugins/body-transformer.lua [hunk 2]
```diff
@@ -108,6 +112,12 @@ local decoders = {
     json = function(data)
         return core.json.decode(data)
     end,
+    encoded = function(data)
+        return  decode_args(data)
```
Comment (monkeyDluffy6017) [inline apisix/plugins/body-transformer.lua:116]: one indent is required

File: apisix/plugins/body-transformer.lua [hunk 3]
```diff
@@ -108,6 +112,12 @@ local decoders = {
...
+    end,
+    args = function()
+        return  req_get_uri_args()
```
Comment (monkeyDluffy6017) [inline apisix/plugins/body-transformer.lua:119]: ditto

File: apisix/plugins/body-transformer.lua [hunk 4]
```diff
@@ -159,24 +171,30 @@ local function transform(conf, body, typ, ctx)
...
+local function set_input_format(conf, typ, ct, method)
+    if method == "GET" then
+        conf[typ].input_format = "args"
+    end
     if conf[typ].input_format == nil and ct then
         if ct:find("text/xml") then
             conf[typ].input_format = "xml"
         elseif ct:find("application/json") then
             conf[typ].input_format = "json"
+        elseif str_find(ct:lower(), "application/x-www-form-urlencoded", nil, true) then
```
Comment (monkeyDluffy6017) [inline apisix/plugins/body-transformer.lua:183]: Use `core.string.find` instead

File: t/plugin/body-transformer.t [hunk 5]
```diff
@@ -894,3 +894,178 @@ location /demo {
...
+            local opt = {method = "GET"}
+            local httpc = http.new()
+            local res = httpc:request_uri(uri, opt)
+            assert(res.status == 200)
+        }
+    }
+
+
+
+=== TEST 15:  The template uses x-www-form-urlencoded format
```
Comment (monkeyDluffy6017) [inline t/plugin/body-transformer.t:1015]: Wrong title?  test `template_is_base64` here?

File: t/plugin/body-transformer.t [hunk 6]
```diff
@@ -894,3 +894,178 @@ location /demo {
...
+
+            local core = require("apisix.core")
+            local http = require("resty.http")
+            local uri = "http://127.0.0.1:" .. ngx.var.server_port .. "/foobar"
+            local data = {name = "hello", age = 20}
+            local body = ngx.encode_args(data)
+            local opt = {method = "POST", body = body, headers = {["Content-Type"] = "application/x-www-form-urlencoded"}}
+            local httpc = http.new()
+            local res = httpc:request_uri(uri, opt)
+        assert(res.status == 200)
```
Comment (monkeyDluffy6017) [inline t/plugin/body-transformer.t:1069]: Wrong indents

File: t/plugin/body-transformer.t (hunk 1 above)
Comment (monkeyDluffy6017) [reply t/plugin/body-transformer.t:949]: The body is `ngx.encode_args` above

**Comment by monkeyDluffy6017** [comment]
The ci is unstable, i will approve this after fix the ci

**Comment by monkeyDluffy6017** [comment]
@yongxiaodong please merge the master to fix the ci

**Comment by yongxiaodong** [comment]
CI still has errors

**Comment by yongxiaodong** [comment]
done

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by shreemaan-abhishek — APPROVED** [review]

**Pull request merged**